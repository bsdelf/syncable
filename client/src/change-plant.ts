import {Change, ChangePlant, Dict, Resource} from '@syncable/core';

import {ClientContext} from './context';

export abstract class ClientChangePlant extends ChangePlant {
  constructor(public context: ClientContext) {
    super();
  }

  apply(change: Change): void {
    let refDict = change.refs;
    // tslint:disable-next-line:no-null-keyword
    let resourceDict: Dict<Resource> = Object.create(null);

    for (let name of Object.keys(refDict)) {
      let ref = refDict[name];
      let resource = this.context.get(ref);

      if (!resource) {
        throw new Error(`Missing resource [${ref.type}] ${ref.id}`);
      }
    }

    this.process(resourceDict, change);
  }
}
