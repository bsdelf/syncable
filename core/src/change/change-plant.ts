import * as DeepDiff from 'deep-diff';
import _ from 'lodash';
import {Dict, KeyOfValueWithType, ValueWithType} from 'tslang';

import {AccessRight} from '../access-control';
import {Context, ISyncableObjectProvider} from '../context';
import {
  AbstractSyncableObject,
  ISyncable,
  ISyncableObject,
  IUserSyncableObject,
  SyncableId,
  SyncableRef,
  SyncableType,
} from '../syncable';

import {
  ChangePacket,
  ChangePacketId,
  GeneralChange,
  IChange,
  SyncableCreationRef,
} from './change';

export type RefDictToObjectOrCreationRefDict<
  T extends object
> = T extends object
  ? {
      [K in KeyOfValueWithType<Required<T>, SyncableRef>]: NonNullable<
        T[K]
      > extends SyncableRef<infer TSyncableObject>
        ?
            | (NonNullable<T[K]> extends SyncableCreationRef<TSyncableObject>
                ? T[K]
                : TSyncableObject)
            | (undefined extends T[K] ? undefined : never)
        : never
    }
  : never;

export type ChangeToObjectOrCreationRefDict<
  T extends IChange
> = T extends IChange<string, infer TRefDict>
  ? RefDictToObjectOrCreationRefDict<TRefDict>
  : never;

export type RefDictToSyncableDict<T extends object> = T extends object
  ? {
      [K in KeyOfValueWithType<Required<T>, SyncableRef>]: NonNullable<
        T[K]
      > extends SyncableRef<infer TSyncableObject>
        ? NonNullable<T[K]> extends SyncableCreationRef<TSyncableObject>
          ? never
          :
              | TSyncableObject['syncable']
              | (undefined extends T[K] ? undefined : never)
        : never
    }
  : never;

export type ChangeToSyncableDict<T extends IChange> = T extends IChange<
  string,
  infer TRefDict
>
  ? RefDictToSyncableDict<TRefDict>
  : never;

export type ChangeToSyncable<T extends IChange> = T extends IChange<
  string,
  infer TRefDict
>
  ? NonNullable<ValueWithType<RefDictToSyncableDict<TRefDict>, any>>
  : never;

export type RefDictToCreation<T extends object> = ValueWithType<
  T,
  SyncableCreationRef
>;

export type ChangeToCreation<T extends IChange> = T extends IChange<
  string,
  infer TRefDict
>
  ? SyncableType<ValueWithType<Required<TRefDict>, SyncableCreationRef>>
  : never;

export interface ChangePlantProcessingResultUpdateItem {
  diffs: deepDiff.IDiff[];
  snapshot: ISyncable;
}

export interface ChangePlantProcessingResult {
  id: ChangePacketId;
  updates: Dict<ChangePlantProcessingResultUpdateItem>;
  creations: ISyncable[];
  removals: SyncableRef[];
  notificationPacket: NotificationPacket | undefined;
}

export interface ChangePlantProcessingResultWithTimestamp
  extends ChangePlantProcessingResult {
  timestamp: number;
}

export type ChangePlantProcessorCreateOperation<TChange extends IChange> = (
  creation: ChangeToCreation<TChange>,
) => void;

export type ChangePlantProcessorRemoveOperation<TChange extends IChange> = (
  removal:
    | Extract<keyof ChangeToSyncableDict<TChange>, string>
    | ChangeToSyncable<TChange>,
) => void;

export interface INotification {
  type: string;
}

export type NotificationPacket<
  TNotification extends INotification = INotification
> = TNotification & {id: ChangePacketId};

export type ChangePlantProcessorNotifyOperation<
  TNotification extends INotification = INotification
> = (notification: TNotification) => void;

export interface ChangePlantProcessorExtra<
  TUser extends IUserSyncableObject = IUserSyncableObject,
  TChange extends IChange = GeneralChange
> {
  context: Context<TUser>;
  options: TChange['options'];
  create: ChangePlantProcessorCreateOperation<TChange>;
  remove: ChangePlantProcessorRemoveOperation<TChange>;
  notify: ChangePlantProcessorNotifyOperation;
}

export type ChangePlantProcessor<
  TUser extends IUserSyncableObject = IUserSyncableObject,
  TChange extends IChange = GeneralChange
> = (
  syncables: ChangeToSyncableDict<TChange>,
  objects: ChangeToObjectOrCreationRefDict<TChange>,
  data: ChangePlantProcessorExtra<TUser, TChange>,
) => void;

export type ChangePlantBlueprint<
  TUser extends IUserSyncableObject,
  TChange extends IChange
> = {
  [K in TChange['type']]: ChangePlantProcessor<
    TUser,
    Extract<TChange, {type: K}>
  >
};

export interface ChangePlantGenericParams {
  user: IUserSyncableObject;
  change: IChange;
  notification: INotification;
}

interface DefaultChangePlantGenericParams extends ChangePlantGenericParams {
  change: GeneralChange;
}

export class ChangePlant<
  TGenericParams extends ChangePlantGenericParams = DefaultChangePlantGenericParams
> {
  constructor(
    private blueprint: ChangePlantBlueprint<
      TGenericParams['user'],
      TGenericParams['change']
    >,
    private provider: ISyncableObjectProvider,
  ) {}

  process(
    packet: ChangePacket,
    syncableObjectOrCreationRefDict: Dict<
      ISyncableObject | SyncableCreationRef
    >,
    context: Context<TGenericParams['user']>,
  ): ChangePlantProcessingResult;
  process(
    packet: ChangePacket,
    syncableObjectOrCreationRefDict: Dict<
      ISyncableObject | SyncableCreationRef
    >,
    context: Context<TGenericParams['user']>,
    timestamp: number,
  ): ChangePlantProcessingResultWithTimestamp;
  process(
    {id, type, options}: ChangePacket,
    syncableObjectOrCreationRefDict: Dict<
      ISyncableObject | SyncableCreationRef
    >,
    context: Context,
    timestamp?: number,
  ): ChangePlantProcessingResult | ChangePlantProcessingResultWithTimestamp {
    let processor = (this.blueprint as any)[type] as ChangePlantProcessor<
      TGenericParams['user'],
      IChange
    >;

    let provider = this.provider;

    let syncableObjectEntries = Array.from(
      Object.entries(syncableObjectOrCreationRefDict),
    ).filter(
      (
        entry: [string, ISyncableObject | SyncableCreationRef],
      ): entry is [string, ISyncableObject] => {
        let [, object] = entry;
        return object instanceof AbstractSyncableObject;
      },
    );

    let syncableKeys = syncableObjectEntries.map(([key]) => key);

    let syncableDict = syncableObjectEntries.reduce(
      (dict, [name, object]) => {
        dict[name] = object.syncable;
        return dict;
      },
      {} as Dict<ISyncable>,
    );

    let syncableObjectMap = new Map<SyncableId, ISyncableObject>();

    let syncableObjectDict = syncableObjectEntries.reduce(
      (dict, [name, object]) => {
        syncableObjectMap.set(object.syncable._id, object);

        dict[name] = object;

        return dict;
      },
      {} as Dict<ISyncableObject>,
    );

    let clonedSyncableDict = _.mapValues(syncableDict, syncable =>
      _.cloneDeep(syncable),
    );

    let creations: ISyncable[] = [];
    let removals: SyncableRef[] = [];
    let notificationPacket:
      | NotificationPacket<TGenericParams['notification']>
      | undefined;

    let create: ChangePlantProcessorCreateOperation<
      GeneralChange
    > = creation => {
      let _creation = creation as ISyncable;

      if (timestamp !== undefined) {
        _creation._timestamp = timestamp;
      }

      creations.push(_creation);
    };

    let remove: ChangePlantProcessorRemoveOperation<
      GeneralChange
    > = removal => {
      let object;

      if (typeof removal === 'string') {
        object = syncableObjectDict[removal];
      } else {
        object = syncableObjectMap.get(removal._id)!;
      }

      object.validateAccessRights(['full'], context);

      removals.push(object.ref);
    };

    let notify: ChangePlantProcessorNotifyOperation<
      TGenericParams['notification']
    > = notification => {
      notificationPacket = {
        id,
        ...(notification as INotification),
      };
    };

    processor(
      clonedSyncableDict,
      syncableObjectOrCreationRefDict as ChangeToObjectOrCreationRefDict<
        GeneralChange
      >,
      {
        context,
        options,
        create: create as ChangePlantProcessorCreateOperation<
          TGenericParams['change']
        >,
        remove: remove as ChangePlantProcessorRemoveOperation<
          TGenericParams['change']
        >,
        notify,
      } as ChangePlantProcessorExtra<
        TGenericParams['user'],
        TGenericParams['change']
      >,
    );

    let updateDict: Dict<ChangePlantProcessingResultUpdateItem> = {};

    for (let key of syncableKeys) {
      let latestSyncable = syncableDict[key];
      let updatedSyncableClone = clonedSyncableDict[key];

      if (timestamp !== undefined) {
        updatedSyncableClone._timestamp = timestamp;
      }

      let diffs = DeepDiff.diff(latestSyncable, updatedSyncableClone);

      if (
        !diffs ||
        !diffs.length ||
        (diffs.length === 1 && diffs[0].path[0] === '_timestamp')
      ) {
        continue;
      }

      let requiredRightSet = new Set<AccessRight>();

      let latestAssociations = provider
        .resolveAssociations(latestSyncable)
        .filter(association => association.secures);
      let updatedAssociations = provider
        .resolveAssociations(updatedSyncableClone)
        .filter(association => association.secures);

      let securingAssociationChanged = !!_.xorBy(
        latestAssociations,
        updatedAssociations,
        ({ref: {type, id}}) => `${type}-${id}`,
      ).length;

      if (securingAssociationChanged) {
        requiredRightSet.add('full');
      }

      for (let diff of diffs) {
        let propertyName = diff.path[0];

        if (
          propertyName === '_id' ||
          propertyName === '_type' ||
          propertyName === '_extends'
        ) {
          throw new Error('Invalid operation');
        }

        if (/^_(?!timestamp)$/.test(propertyName)) {
          requiredRightSet.add('full');
        } else {
          requiredRightSet.add('write');
        }
      }

      if (requiredRightSet.size) {
        syncableObjectDict[key].validateAccessRights(
          Array.from(requiredRightSet),
          context,
        );
      }

      updateDict[key] = {diffs, snapshot: updatedSyncableClone};
    }

    return {
      id,
      timestamp,
      updates: updateDict,
      creations: creations || [],
      removals: removals || [],
      notificationPacket,
    };
  }
}
