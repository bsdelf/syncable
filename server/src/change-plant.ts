import {Change, ChangePlant, Dict, Resource, ResourceRef} from '@syncable/core';
import * as v from 'villa';

import {ServerContext} from './context';

export abstract class ServerChangePlant extends ChangePlant {
  constructor(public context: ServerContext) {
    super();
  }

  async apply(change: Change): Promise<void> {
    let refDict = change.refs;
    // tslint:disable-next-line:no-null-keyword
    let resourceDict: Dict<Resource> = Object.create(null);

    let entries = Object.keys(refDict).map((name): [string, ResourceRef] => [
      name,
      refDict[name],
    ]);

    await v.parallel(entries, async ([name, ref]) => {
      resourceDict[name] = await this.context.resolve(ref);
    });

    this.process(resourceDict, change);
  }
}
