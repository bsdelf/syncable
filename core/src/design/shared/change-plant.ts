import {Change, ChangePlant, ChangePlantBlueprint} from '../../change';
import {Syncable, SyncableRef} from '../../syncable';
import {TagSyncable} from './tag';

type AllChange = TagChange;

interface TagChangeOptions {
  foo: boolean;
}

interface TagChangeRefDict {
  target: SyncableRef;
  tag: SyncableRef<TagSyncable>;
}

interface TagChange extends Change<'tag', TagChangeRefDict, TagChangeOptions> {}

const blueprint: ChangePlantBlueprint<AllChange> = {
  tag({target, tag}, options): void {},
};

const changePlant = new ChangePlant(blueprint);

let change!: TagChange;

changePlant.process(change, {
  target: {} as Syncable,
  tag: {} as TagSyncable,
});
