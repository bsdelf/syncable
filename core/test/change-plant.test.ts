import {
  ChangePacketUID,
  ChangePlant,
  Context,
  SyncableCreationRef,
  createSyncable,
  getSyncableRef,
} from '../bld';

import {Task, TaskId, User, changePlantBlueprint} from './change-plant';

let context = new Context<User>('server');

let plant = new ChangePlant(changePlantBlueprint);

let taskSyncable = createSyncable<Task>(
  {
    creation: true,
    type: 'task',
    id: 'task-id' as TaskId,
  },
  {
    brief: 'hello, world.',
  },
);

let task = new Task(taskSyncable);

test('should update task brief', () => {
  let result = plant.process(
    {
      uid: 'change-packet-id' as ChangePacketUID,
      type: 'task:update-brief',
      refs: {
        task: getSyncableRef(taskSyncable),
      },
      options: {
        brief: 'hello, jest.',
      },
    },
    {
      task,
    },
    context,
  );

  expect(result).toMatchSnapshot();
});

test('should create task', () => {
  let taskRef: SyncableCreationRef<Task> = {
    creation: true,
    type: 'task',
    id: 'task-id' as TaskId,
  };

  let result = plant.process(
    {
      uid: 'change-packet-id' as ChangePacketUID,
      type: 'task:create',
      refs: {
        task: taskRef,
      },
      options: {
        brief: 'hello, create.',
      },
    },
    {
      task: taskRef,
    },
    context,
  );

  expect(result).toMatchSnapshot();
});

test('should remove task', () => {
  let result = plant.process(
    {
      uid: 'change-packet-id' as ChangePacketUID,
      type: 'task:remove',
      refs: {
        task: getSyncableRef(taskSyncable),
      },
      options: {},
    },
    {
      task,
    },
    context,
  );

  expect(result).toMatchSnapshot();
});
