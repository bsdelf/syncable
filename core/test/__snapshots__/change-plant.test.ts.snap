// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should create task 1`] = `
Object {
  "creations": Array [
    Object {
      "_id": "task-id",
      "_timestamp": 0,
      "_type": "task",
      "brief": "hello, create.",
    },
  ],
  "removals": Array [],
  "timestamp": undefined,
  "uid": "change-packet-id",
  "updates": Object {},
}
`;

exports[`should remove task 1`] = `
Object {
  "creations": Array [],
  "removals": Array [
    Object {
      "id": "task-id",
      "type": "task",
    },
  ],
  "timestamp": undefined,
  "uid": "change-packet-id",
  "updates": Object {},
}
`;

exports[`should update task brief 1`] = `
Object {
  "creations": Array [],
  "removals": Array [],
  "timestamp": undefined,
  "uid": "change-packet-id",
  "updates": Object {
    "task": Object {
      "diffs": Array [
        DiffEdit {
          "kind": "E",
          "lhs": "hello, world.",
          "path": Array [
            "brief",
          ],
          "rhs": "hello, jest.",
        },
      ],
      "snapshot": Object {
        "_id": "task-id",
        "_timestamp": 0,
        "_type": "task",
        "brief": "hello, jest.",
      },
    },
  },
}
`;
